File: RUST/sfo_24_reverse_a_linked_list_s2/src/main.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}
//
impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub struct Solution {}
pub struct Utils {}

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Solution::recur(head)
    }
}

impl Utils {
    pub fn vec_to_link_list(num: Vec<i32>) -> Option<Box<ListNode>> {
        let mut head = ListNode::new(0);
        let mut current = &mut head;
        for &num in num.iter() {
            current.next = Some(Box::new(ListNode::new(num)));
            current = current.next.as_mut().unwrap();
        }
        head.next
    }

    pub fn print_link_list(head: &Option<Box<ListNode>>) {
        let mut current = head;
        while let Some(node) = current {
            print!("{}->", node.val);
            current = &node.next;
        }
        println!("NULL");
    }
}

fn main() {
    let input = vec![1, 2, 3, 4, 5];
    let input = Utils::vec_to_link_list(input);
    Utils::print_link_list(&input);
    let ans = Solution::reverse_list(input);
    Utils::print_link_list(&ans);
}

File: CPP/sfo_24_reverse_a_linked_list_s2/main.cc
#include "../include/include.hpp"

class Solution {
public:
    ListNode* reverseList( ListNode* head ) {
        return recur( head, NULL );
    }

    ListNode* recur( ListNode* cur, ListNode* pre ) {
        if ( cur == NULL ) return pre;
        ListNode* res = recur( cur->next, cur );
        cur->next     = pre;
        return res;
    }
};

int main( int argc, char* argv[] ) {
    ListNode* head = vectorToLinkedList( vector<int>{ 1, 2, 3, 4, 5 } );
    Solution* stl  = new Solution();
    ListNode* res  = stl->reverseList( head );
    PrintUtil::printLinkedList( res );
}
